### DISCUSSION TITLE
Simulating a Pauli tensor faster than as a generalised permutation matrix

### DISCUSSION THREAD

**Discussion by TysonRayJones** [NVIDIA/cuQuantum#154] [Ideas]
Hi there,

Imagine that we seek to _apply_ a tensor of Pauli operators (rather than just calculate their expectation values). That is, they appear as same-column gates in our circuit like might appear due to twirling, or Monte Carlo decomposition of Pauli channels, etc.

<img width="42" alt="image" src="https://github.com/user-attachments/assets/0789275f-4375-4366-9b86-cb44a30195ce">

We could inefficiently effect these in-turn as one-qubit gates. It turns out that's a factor `#paulis` _slower_ than necessary, and is especially a shame when the operators have common controls:
<img width="97" alt="image" src="https://github.com/user-attachments/assets/70a45fb8-0351-40a3-9bcf-209a4ae8a029">

It is ideal to simulate them as a _single_ operator:

<img width="48" alt="image" src="https://github.com/user-attachments/assets/0f38b30c-40bd-4ab7-bfd3-48ad66225103">

Tensors of Pauli operators have a special form owed to each being diagonal or anti-diagonal, and can be simulated as quickly as a _single_ Pauli operator (with _no_ necessary matrix allocations) as outlined in Sec IV E [here](https://arxiv.org/pdf/2311.01512). It is my understanding that it is not currently possible to leverage this in cuStateVec.

Their Z-basis matrices are simple, specific instances of _generalised permutation matrices_:

> <img width="397" alt="image" src="https://github.com/user-attachments/assets/c259692d-dce7-4590-8d6e-34491a91107e">

They can ergo be simulated using cuStateVec's [`custatevecApplyGeneralizedPermutationMatrix`](https://docs.nvidia.com/cuda/cuquantum/latest/custatevec/api/functions.html?highlight=pauli#_CPPv443custatevecApplyGeneralizedPermutationMatrix18custatevecHandle_tPv14cudaDataType_tK8uint32_tP17custatevecIndex_tPKv14cudaDataType_tK7int32_tPK7int32_tK8uint32_tPK7int32_tPK7int32_tK8uint32_tPv6size_t), however a few issues remain:
- we must ourselves build the diagonal elements which is a chore (involving counting the parity of Y operators, tracking where Y and Z non-unity coeffs are effected, etc). A similar chore is likely already being automatically performed in [`custatevecApplyPauliRotation`](https://docs.nvidia.com/cuda/cuquantum/latest/custatevec/api/functions.html?highlight=pauli#_CPPv428custatevecApplyPauliRotation18custatevecHandle_tPv14cudaDataType_tK8uint32_tdPK17custatevecPauli_tPK7int32_tK8uint32_tPK7int32_tPK7int32_tK8uint32_t), if it is using an algorithm similar to [Sec IV G](https://arxiv.org/pdf/2311.01512).
-  we must ourselves build the permutation matrices (by tracking X and Y operators, etc). This is a totally superfluous allocation of memory which is exponential in the number of Paulis (I _think_ as bad as `4^#paulis`)! Essentially, the Pauli tensor simply swaps pairs of amplitudes and never moves them "in a loop" like a general permutation matrix might, so the final destination index of the amplitudes is simply and cheaply encoded using a fixed number of bitmasks, rather than an exponentially big matrix.

Ergo a new, dedicated function for simulating Pauli tensors in cuStateVec would be very valuable and _much_ faster than the existing method (linearly faster than in-turn simulation, and exponentially faster than via permutation matrices).
And notably, it will _**be easy**_; the logic for simulating the Pauli tensor is _identical_ to optimal simulation of the Pauli gadget (`custatevecApplyPauliRotation`); they differ only by constant coefficients multiplied onto each amplitude (courtesy of [Euler's formula](https://en.wikipedia.org/wiki/Euler%27s_formula)).

You can view a non-distributed implementation of the algorithm [here](https://github.com/TysonRayJones/Distributed-Full-State-Algorithms/blob/f29bd6b81156bd315d414cd43f94602be9e00617/src/local_statevector.hpp#L102) with arguments prepared [here](https://github.com/TysonRayJones/Distributed-Full-State-Algorithms/blob/f29bd6b81156bd315d414cd43f94602be9e00617/src/distributed_statevector.hpp#L244), which is derived in Sec IV E [here](https://arxiv.org/pdf/2311.01512). It is trivial to introduce control qubits in arbitrary states, which merely eliminate iterations of the main loop (each of which modifies one amplitude).

In my application (which I'll soon have the public code to link), I will have to fall-back to using a custom kernel implementing the above algorithm - it would be fantastic to update it to use a cuStateVec function in the future!

**Comment by ymagchi** [comment]
Hi @TysonRayJones,

Thank you so much for your feedback.
We may be able to lift this requirement in the future, while the cuStateVec library has prioritized the functionality to pass an arbitrary matrix.

I'd like to add to the computation cost of `custatevecApplyGeneralizedPermutationMatrix`: both permutation/diagonal matrices will be passed as arrays of size `2^#Paulis` and I expect that the kernel would be almost memory-bound because each state vector element requires only 1 matrix element during the multiplication. We will need some overhead to create a permutation table but the perf of `custatevecApplyGeneralizedPermutationMatrix` will not slow down exponentially to the number of Paulis.

**Reply by TysonRayJones** [reply]
Ahh right you are! Each updated amplitude uses a _single_ permutation element so the number of operations does _not_ scale exponentially with `#Paulis` - I unconsciously borrowed the logic of the many-target general matrix, my bad!

It is still true however that, in addition to the overhead of creation, querying that exponentially large matrix will have to bite _somewhere_ due to caching. For example, 25 Paulis will mean querying a `permutation` array of 256 MiB in a hot loop/dispatch. If CUDA's caching slowdown was logarithmic with memory size, use of the generalised permutation matrices would still induce a _linear_ slowdown over a bespoke method. Of course we might never reach such a regime before running out of memory ¯\\_(ツ)_/¯
